crates/dns/src/proxy.rs

```diff
@@ -0,0 +1,261 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::resolver::{Answer, Resolver};
+use std::sync::Arc;
+use tracing::{error, warn};
+use trust_dns_proto::op::{Edns, Header, MessageType, OpCode, ResponseCode};
+use trust_dns_proto::rr::Record;
+use trust_dns_resolver::error::ResolveErrorKind;
+use trust_dns_server::authority::{LookupError, MessageResponse, MessageResponseBuilder};
+use trust_dns_server::server::{Request, RequestHandler, ResponseHandler, ResponseInfo};
+
+/// A Trust-DNS [ResponseHandler] that proxies all DNS requests.
+///
+/// A proxy is fundamentally different than an `Authority` in TrustDNS, since the answers may
+/// or may not be authoritative based on whether they are served locally or forwarded. It is
+/// for this reason that we can't implement the proxy using existing TrustDNS server structures
+/// `Catalog` and `Authority`.
+// TODO(nmittler): Consider upstreaming this to TrustDNS
+pub struct Proxy {
+    resolver: Arc<dyn Resolver>,
+}
+
+impl Proxy {
+    /// Creates a new request handler for the resolver.
+    pub fn new(resolver: Arc<dyn Resolver>) -> Self {
+        Self { resolver }
+    }
+
+    async fn lookup<R: ResponseHandler>(
+        &self,
+        request: &Request,
+        response_handle: R,
+    ) -> ResponseInfo {
+        match self.resolver.lookup(request).await {
+            Ok(answer) => send_lookup(request, response_handle, answer).await,
+            Err(e) => send_lookup_error(request, response_handle, e).await,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl RequestHandler for Proxy {
+    async fn handle_request<R: ResponseHandler>(
+        &self,
+        request: &Request,
+        response_handle: R,
+    ) -> ResponseInfo {
+        match request.message_type() {
+            MessageType::Query => match request.op_code() {
+                OpCode::Query => self.lookup(request, response_handle).await,
+                _ => {
+                    warn!("unimplemented op_code: {:?}", request.op_code());
+                    send_error(request, response_handle, ResponseCode::NotImp).await
+                }
+            },
+            MessageType::Response => {
+                warn!("got a response as a request from id: {}", request.id());
+                send_error(request, response_handle, ResponseCode::FormErr).await
+            }
+        }
+    }
+}
+
+async fn send_lookup<R: ResponseHandler>(
+    request: &Request,
+    response_handle: R,
+    answer: Answer,
+) -> ResponseInfo {
+    let mut response_header = Header::response_from_request(request.header());
+
+    // We are the authority here, since we control DNS for known hostnames
+    response_header.set_authoritative(answer.is_authoritative());
```

Comment:
Does this mimic the current Go DNS logic?  Are we going to confuse anything if some responses are marked as authoritative and some not, from what the client (believes is) the same DNS server?

Reply:
The behavior should be the same as the Istio code: https://github.com/istio/istio/blob/5aa2900dca83190962aea2eac8712ca88ca63da3/pkg/dns/client/dns.go#L289